firebase-common/src/main/java/com/google/firebase/concurrent/LimitedConcurrencyExecutor.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.concurrent;
+
+import androidx.annotation.GuardedBy;
+import com.google.firebase.components.Preconditions;
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingQueue;
+
+class LimitedConcurrencyExecutor implements Executor {
+
+  private final Executor delegate;
+
+  @GuardedBy("queue")
+  private int activeCount;
+
+  private final LinkedBlockingQueue<Runnable> queue = new LinkedBlockingQueue<>();
+
+  LimitedConcurrencyExecutor(Executor delegate, int concurrency) {
+    Preconditions.checkArgument(concurrency > 0, "concurrency must be positive.");
+    this.delegate = delegate;
+    activeCount = concurrency;
+  }
+
+  @Override
+  public void execute(Runnable command) {
```

Comment:
In both `LimitedExecutor` and `PausableExecutor` any new `Runnable` that comes is executed first, even if there are tasks waiting in the queue, and only after it finishes the next one in the queue is maybe executed. This can cause in extreme scenarios that tasks that were enqueued when the executors were 'offline' starve because new jobs have priority. Always enqueuing and execute tasks in FIFO order would guarantee fairness.

Reply:
Great point, I will rewrite them to have fifo semanics